insights/client/apps/ansible/playbook_verifier/__init__.py

```diff
@@ -20,18 +36,106 @@ def __str__(self):
+                raise PlaybookVerificationError(message="PUBLIC KEY NOT IMPORTED: Public key not imported from key file")
+
+            return importResults
+    except Exception as e:
+        raise PlaybookVerificationError(message="PUBLIC KEY IMPORT ERROR: {0}".format(e))
+
+
+def excludeDynamicElements(snippet):
+    exclusions = snippet['vars']['insights_signature_exclude'].split(',')
+
+    for element in exclusions:
+        element = element.split('/')
+
+        # remove empty strings
+        element = [string for string in element if string != '']
+
+        if (len(element) == 1 and element[0] in EXCLUDABLE_VARIABLES):
+            del snippet[element[0]]
+        elif (len(element) == 2 and element[0] in EXCLUDABLE_VARIABLES):
+            try:
+                del snippet[element[0]][element[1]]
+            except:
+                raise PlaybookVerificationError(message='INVALID FIELD: the variable {0} defined in insights_signature_exclude does not exist.'.format(element))
+        else:
+            raise PlaybookVerificationError(message='INVALID EXCLUSION: the variable {0} is not a valid exclusion.'.format(element))
+
+    return snippet
+
+
+def executeVerification(snippet, encodedSignature):
+    serializedSnippet = pickle.dumps(snippet)
+    decodedSignature = base64.b64decode(encodedSignature)
+
+    # load public key
+    try:
+        getPublicKey()
+    except PlaybookVerificationError as e:
+        raise e
+
+    fd, fn = tempfile.mkstemp()
+    os.write(fd, decodedSignature)
+    os.close(fd)
+
+    result = gpg.verify_data(fn, serializedSnippet)
+    os.unlink(fn)
+
+    return result
+
+
+def verifyPlaybookSnippet(snippet):
+    encodedSignature = snippet['vars'][SIGKEY]
+    snippetCopy = copy.deepcopy(snippet)
+
+    snippetCopy = excludeDynamicElements(snippetCopy)
+
+    return executeVerification(snippetCopy, encodedSignature)
+
+
+def verify(playbook, checkVersion = True, skipVerify = False):
     """
     Verify the signed playbook.
 
-    Input: stringified "broken" unverified playbook
-    Output: stringified "verified" playbook
-    Error: exception
+    Input: unverified playbook (dictionary format)
+    Output: "verified" playbook (dictionary format)
+    Error: Playbook Verification failure / Playbook Signature not found.
     """
-    # Skeleton implementation ... "bless" the incoming playbook
-    ERROR = os.getenv('ANSIBLE_PLAYBOOK_VERIFIER_THROW_ERROR')
-    if ERROR:
-        raise PlaybookValidationError()
+    logger.info('Playbook Verification has started')
+
+    # Egg Version Check
+    eggVersioningCheck(checkVersion)
+
+    if not skipVerify:
+        for snippet in playbook:
+            if ('vars' not in snippet.keys()):
+                raise PlaybookVerificationError(message='VARS FIELD NOT FOUND: Verification failed')
+            elif (SIGKEY not in snippet['vars']):
+                raise PlaybookVerificationError(message='SIGNATURE NOT FOUND: Verification failed')
```

Comment:
I'm possibly being nit-picky here...forgive me.  :)

It feels like this logic belongs in the verifyPlaybookSnippet() function.  The verify() method appears to be "borrowing" this logic.

Reply:
done!